dbms/src/AggregateFunctions/AggregateFunctionSumMap.h

```diff
@@ -0,0 +1,189 @@
+        const auto & keys = field_keys.get<Array &>();
+
+        Field field_values;
+        column_values.get(row_num, field_values);
+        const auto & values = field_values.get<Array &>();
+
+        auto & merged_maps = this->data(place).merged_maps;
+
+        if (keys.size() != values.size())
+            throw Exception("Sizes of keys and values arrays do not match", ErrorCodes::LOGICAL_ERROR);
+
+        size_t size = keys.size();
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (merged_maps.find(keys[i]) != merged_maps.end())
+                merged_maps[keys[i]] += values[i];
+            else
+                merged_maps[keys[i]] = values[i];
+        }
+    }
+
+    void merge(AggregateDataPtr place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+        const auto & rhs_maps = this->data(rhs).merged_maps;
+
+        for (const auto &rhs_map : rhs_maps)
+        {
+            if (merged_maps.find(rhs_map.first) != merged_maps.end())
+                merged_maps[rhs_map.first] += rhs_map.second;
+            else
+                merged_maps[rhs_map.first] = rhs_map.second;
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+    {
+        /// Serialize merged_maps as two vectors. Using boost::archive could be better but it's unavailable.
+        const auto & merged_maps = this->data(place).merged_maps;
+        size_t size = merged_maps.size();
+
+        Array keys, values;
+        keys.reserve(size);
+        values.reserve(size);
+        for (const auto &v : merged_maps)
+        {
+            keys.push_back(v.first);
+            values.push_back(v.second);
+        }
+
+        writeVarUInt(size, buf);
+        buf.write(reinterpret_cast<const char *>(&keys[0]), size * sizeof(keys[0]));
+        buf.write(reinterpret_cast<const char *>(&values[0]), size * sizeof(values[0]));
+    }
+
+    void deserialize(AggregateDataPtr place, ReadBuffer & buf, Arena *) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+
+        size_t size = 0;
+        readVarUInt(size, buf);
+
+        Array keys, values;
+        keys.resize(size);
+        values.resize(size);
+        buf.read(reinterpret_cast<char *>(&keys[0]), size * sizeof(keys[0]));
+        buf.read(reinterpret_cast<char *>(&values[0]), size * sizeof(values[0]));
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            merged_maps[keys[i]] = values[i];
+        }
+    }
+
+    void insertResultInto(ConstAggregateDataPtr place, IColumn & to) const override
+    {
+        auto & to_array = static_cast<ColumnArray &>(to);
+        auto & to_data = to_array.getData();
+        auto & to_offsets = to_array.getOffsets();
+
+        const auto & merged_maps = this->data(place).merged_maps;
+        size_t size = merged_maps.size();
+
+        Array keys, values;
```

Comment:
Excessive copy. Better to insert into nested column directly.

Reply:
I've tried to do that but couldn't access sub-arrays of `ColumnArray`. I also tried to find similar implementation with usage of `getNestedColumn` but existing examples are quite different. There is no `NestedColumn` column type.

Can you please explain bit more how to implement the suggested change?

Reply:
Return type is `Array(Array(T))`

`auto & to_data = to_array.getData();`

`to_data` is elements of result arrays, so `Array(T)`.

Now if you get `static_cast<DataTypeArray &>(to_data).getData()`, you will get column of type `T` that will contain all elements of all arrays.

And you can insert individual elements into it, one by one.

Reply:
PS. Why return type is not `Tuple(Array(K), Array(V))`?
This will allow different types of keys and values.
For example, keys can be strings and values - integers.
And manipulation with the result will be more simple.

Better to change signature before merge.